baselines/mnist/variational_inference.py

```diff
@@ -0,0 +1,474 @@
+
+  def log_likelihood(y_true, y_sample):
+    """Expected conditional log-likelihood."""
+    del y_sample  # unused arg
+    return model.output.distribution.log_prob(tf.squeeze(y_true))
+
+  def kl(y_true, y_sample):
+    """KL-divergence."""
+    del y_true  # unused arg
+    del y_sample  # unused arg
+    sampling_cost = sum(
+        [l.kl_cost_weight + l.kl_cost_bias for l in model.layers])
+    return sum(model.losses) * n_train + sampling_cost
+
+  def elbo(y_true, y_sample):
+    return log_likelihood(y_true, y_sample) * n_train - kl(y_true, y_sample)
+
+  return negative_log_likelihood, accuracy, log_likelihood, kl, elbo
+
+
+def main(argv):
+  del argv  # unused arg
+  np.random.seed(FLAGS.seed)
+  tf.random.set_seed(FLAGS.seed)
+  tf.io.gfile.makedirs(FLAGS.output_dir)
+  tf1.disable_v2_behavior()
+
+  session = tf1.Session()
+  with session.as_default():
+    x_train, y_train, x_test, y_test = utils.load(FLAGS.dataset, session)
+    n_train = x_train.shape[0]
+
+    num_classes = int(np.amax(y_train)) + 1
+    model = lenet5(n_train, x_train.shape[1:], num_classes)
+    for l in model.layers:
+      l.kl_cost_weight = l.add_weight(
+          name='kl_cost_weight',
+          shape=(),
+          initializer=tf.constant_initializer(0.),
+          trainable=False)
+      l.kl_cost_bias = l.add_variable(
+          name='kl_cost_bias',
+          shape=(),
+          initializer=tf.constant_initializer(0.),
+          trainable=False)
+
+    [negative_log_likelihood,
+     accuracy,
+     log_likelihood,
+     kl,
+     elbo] = get_losses_and_metrics(model, n_train)
+    metrics = [elbo, log_likelihood, kl, accuracy]
+    tensorboard = tf1.keras.callbacks.TensorBoard(
+        log_dir=FLAGS.output_dir,
+        update_freq=FLAGS.batch_size * FLAGS.validation_freq)
+
+    def fit_fn(model,
+               steps,
+               initial_epoch):
+      return model.fit(
+          x=x_train,
+          y=y_train,
+          batch_size=FLAGS.batch_size,
+          epochs=initial_epoch + (FLAGS.batch_size * steps) // n_train,
+          initial_epoch=initial_epoch,
+          validation_data=(x_test, y_test),
+          validation_freq=max(
+              (FLAGS.validation_freq * FLAGS.batch_size) // n_train, 1),
+          verbose=1,
+          callbacks=[tensorboard])
+
+    model.compile(
+        optimizer=tf.keras.optimizers.Adam(lr=float(FLAGS.learning_rate)),
+        loss=negative_log_likelihood,
+        metrics=metrics)
+    session.run(tf1.initialize_all_variables())
+
+    train_epochs = (FLAGS.training_steps * FLAGS.batch_size) // n_train
+    fit_fn(model, FLAGS.training_steps, initial_epoch=0)
+
+    labels = tf.keras.layers.Input(shape=y_train.shape[1:])
+    ll = tf.keras.backend.function([model.input, labels], [
+        model.output.distribution.log_prob(tf.squeeze(labels)),
+        model.output.distribution.logits
+    ])
+
+    base_metrics = [
+        utils.ensemble_metrics(x_train, y_train, model, ll),
```

Comment:
ensemble_metrics by default only draws one sample from the posterior. This should be utils.ensemble_metrics(x_train, y_train, model, ll, n_samples=10). Could be the culprit for the performance disparity.

Reply:
same in line 386, 401, 402

Reply:
Done.

Reply:
Update: re-ran; results did not change too much. Will look into tuning after TF2 conversions.